text/0225-configuration.md

```diff
@@ -0,0 +1,410 @@
+  * Property value types will not change, except that integers MAY become floating points.
+  * No additional required properties.
+
+Allowable changes:
+
+* For major versions: All changes are permitted.
+* For minor versions:
+  * Addition of optional properties.
+  * Required property keys may become optional.
+  * Removal of properties, provided that the property key is not reused in the future.
+
+SDKs validating configuration *MUST* fail when they encounter a configuration with an unsupported version. Generally, this means fail when encountering a major version which is not recognized. An SDK might choose to maintain a library of validators / parsers for each major version, and use the configuration version to select and use the correct instance. Differences in minor versions (except pre-1.0 minor versions) SHOULD be acceptable, with the caveat that allowable property additions and removals MAY result in configuration that is different than excepted.
+
+## Trade-offs and mitigations
+
+### Additional method to configure OpenTelemetry
+
+If the implementation suggested in this OTEP goes ahead, users will be presented with another mechanism for configuring OpenTelemetry. This may cause confusion for users who are new to the project. It may be possible to mitigate the confusion by providing users with best practices and documentation.
+
+### Many ways to configure may result in users not knowing what is configured
+
+As there are multiple mechanisms for configuration, it's possible that the active configuration isn't what was expected. This could happen today, and one way it could be mitigated would be by providing a mechanism to list the active OpenTelemetry configuration.
+
+### Errors or difficulty in configuration files
+
+Configuration files provide an opportunity for misconfiguration. A way to mitigate this would be to provide clear messaging and fail quickly when misconfiguration occurs.
+
+## Prior art and alternatives
+
+The working group looked to the OpenTelemetry Collector and OpenTelemetry Operator for inspiration and guidance.
+
+### Alternative schema languages
+
+In choosing to recommend JSON schema, the working group looked at the following options:
+
+* [Cue](https://cuelang.org/) - A promising simpler language to define a schema, the working group decided against CUE because:
+  * Tooling available for validating CUE files in languages outside of Go were limited.
+  * Familiarity and learning curve would create problems for both users and contributors of OpenTelemetry.
+* [Protobuf](https://developers.google.com/protocol-buffers) - With protobuf already used heavily in OpenTelemetry, the format was worth investigating as an option to define the schema. The working group decided against Protobuf because:
+  * Validation errors are the result of serlization errors which can be difficult to interpret.
+  * Limitations in the schema definition language result in poor ergonomics if type safety is to be retained.
```

Comment:
I would like to see more concrete examples of these restrictions.

These reasons only consider the client/consumption side. To me the remote feature of configuration is one of very important ones, and that means we now have to introduce a whole new tool chain of JSON Schema in the server-side authoring & serving of the configs. I would rather stick with Protobuf, to have a single way to manage data models.

Reply:
I took a stab at modeling a config file with protobuf (schema [here](https://github.com/MrAlias/otel-schema/blob/main/protobuf/schema/sdk.proto), example config [here](https://github.com/MrAlias/otel-schema/blob/main/protobuf/kitchen-sink.yaml)). Sometimes protobuf works great, but in several circumstances we'd be forced to have worse ergonomics to work around protobuf limitations. 

One example is trying to define a set of resource attributes. With json schema it can look like:

```
sdk:
  resource:
    attributes:
      service.name: my-service
```

With protobuf the same has to look like:

```
sdk:
  resource:
    attributes:
      service.name:
        string_value: my-service
```

Despite being able to infer type of `service.name` unambiguously, protobuf forces the user to provide additional type information in the keys. 

Another example can be seen in configuring components like exporters, where we'd ideally like to be able to enforce a schema based on the name of the key. 

```
sdk:
  tracer_provider:
    exporters:
      otlp:
        protocol: http/protobuf
        endpoint: http://localhost:4318/v1/traces
      otlp/foo:
        protocol: grpc
        endpoint: http://some-other-host:4317
      zipkin:
        endpoint: http://localhost:9411/api/v2/spans
```

So when key matches `otlp(.*)`, enforce the schema for OTLP exporter. When key matches `zipkin(.*)`, enforce the schema for zipkin.

In protobuf if we want to enforce schemas of exporters, we need to work around the constraints of protobuf in a way that is very apparent to the user. There are different ways to do it, but you might end up with something like:

```
sdk:
  tracer_provider:
    exporters:
      otlp:
        otlp_args:
          protocol: http/protobuf
          endpoint: http://localhost:4318/v1/traces
      otlp/foo:
          otlp_args:
            protocol: grpc
            endpoint: http://some-other-host:4317
      zipkin:
        zipkin_args:
           endpoint: http://localhost:9411/api/v2/spans
```

In this example the keys identify the names of the exporter and are referenced when configuring processors. The values are a generic exporter message type, with `oneof` fields that enforce the schema of the specific exporter. 

To me, the benefit of protobuf is that we get to take advantage of the schema evolution rules we've already developed for otlp. Language SIGs are also familiar with working protobuf, so choosing it alleviates the need to learn a new tech. But the downside is that the files become more awkward to write for users. I was personally I big fan of protobuf, but came around to json schema because of the improved user experience. 

> I would rather stick with Protobuf, to have a single way to manage data models.

There's precedent for json schema in the [semantic convention schema definition](https://github.com/open-telemetry/build-tools/blob/v0.15.1/semantic-conventions/syntax.md#json-schema).

There are few other places we define data models which don't use protobuf:

- Collector configuration
- Telemetry schema [file format](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/schemas/file_format_v1.1.0.md)

Reply:
@yurishkuro do you have any additional thoughts on this?

Reply:
sgtm